File: backend/app/schemas/leaderboard.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.common import Paginated


class LeaderboardEntry(BaseModel):
    rank: int
    user_id: int
    username: str
    initial_balance: float
    portfolio_value: float
    profit_loss: float
    performance_pct: float
    is_current_user: bool = False


class LeaderboardResponse(Paginated[LeaderboardEntry]):
    total_players: int
    computed_at: datetime


class MyRankResponse(BaseModel):
    rank: int | None
    total_players: int
    portfolio_value: float
    profit_loss: float
    performance_pct: float

File: frontend/src/types/leaderboard.ts
import type { Paginated } from './api';

export interface LeaderboardEntry {
  rank: number;
  user_id: number;
  username: string;
  initial_balance: number;
  portfolio_value: number;
  profit_loss: number;
  performance_pct: number;
  is_current_user: boolean;
}

export interface LeaderboardResponse extends Paginated<LeaderboardEntry> {
  total_players: number;
  computed_at: string;
}

export interface MyRank {
  rank: number | null;
  total_players: number;
  portfolio_value: number;
  profit_loss: number;
  performance_pct: number;
}
